gcc/config/riscv/rvp.md

```diff
@@ -0,0 +1,6987 @@
+  "TARGET_ZPN"
+  "kabsw\t%0, %1"
+  [(set_attr "type" "dsp")
+   (set_attr "mode" "SI")])
+
+(define_insn "kabs<mode>2"
+  [(set (match_operand:VECI 0 "register_operand"              "=r")
+	(ss_abs:VECI (match_operand:VECI 1 "register_operand" " r")))]
+  "TARGET_ZPN"
+  "kabs<bits>\t%0, %1"
+  [(set_attr "type"   "simd")
+   (set_attr "mode" "<MODE>")])
+
+;; sub/adddi spn
+(define_insn "*subdi3_rvp"
+  [(set (match_operand:DI 0 "register_operand"               "=r")
+	(minus:DI (match_operand:DI 1 "register_operand" " r")
+		 (match_operand:DI 2 "arith_operand"   " r")))]
+  "TARGET_ZPSF && !TARGET_64BIT"
+{
+  HOST_WIDE_INT tmp;
+  if (CONST_INT_P (operands[2]))
+    {
+      tmp = INTVAL (operands[2]);
+      operands[2] = gen_reg_rtx (DImode);
+      riscv_move_integer (operands[2], operands[2], tmp,
+              DImode, FALSE);
+    }
+  return "sub64 %0, %1, %2";
+}
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+(define_insn "*adddi3_rvp"
+  [(set (match_operand:DI 0 "register_operand"               "=r")
+	(plus:DI (match_operand:DI 1 "register_operand" " r")
+		 (match_operand:DI 2 "arith_operand"   " r")))]
+  "TARGET_ZPSF && !TARGET_64BIT"
+{
+  HOST_WIDE_INT tmp;
+  if (CONST_INT_P (operands[2]))
+    {
+      tmp = INTVAL (operands[2]);
+      operands[2] = gen_reg_rtx (DImode);
+      riscv_move_integer (operands[2], operands[2], tmp,
+              DImode, FALSE);
+    }
+  return "add64 %0, %1, %2";
+}
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+;; k|(uk)|? add
+(define_insn "<uk>add<mode>3"
+  [(set (match_operand:VECI 0 "register_operand"                "=r")
+	(all_plus:VECI (match_operand:VECI 1 "register_operand" " r")
+		       (match_operand:VECI 2 "register_operand" " r")))]
+  "TARGET_ZPN"
+  "<uk>add<bits>\t%0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "<MODE>")])
+
+;; rv64
+(define_insn "rvp_<uk>adddi3"
+  [(set (match_operand:DI 0 "register_operand"              "=r")
+	(saturation_plus:DI (match_operand:DI 1 "register_operand" " r")
+		     (match_operand:DI 2 "register_operand" " r")))]
+  "TARGET_ZPSF"
+  "<uk>add64 %0, %1, %2"
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+;; k|(uk)|? sub
+(define_insn "<uk>sub<mode>3"
+  [(set (match_operand:VECI 0 "register_operand"                  "=r")
+	(all_minus:VECI (match_operand:VECI 1 "register_operand" " r")
+			(match_operand:VECI 2 "register_operand" " r")))]
+  "TARGET_ZPN"
+  "<uk>sub<bits> %0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "<MODE>")])
+
+;; rv64
+(define_insn "rvp_<uk>subdi3"
+  [(set (match_operand:DI 0 "register_operand"               "=r")
+	(saturation_minus:DI (match_operand:DI 1 "register_operand" " r")
+		      (match_operand:DI 2 "register_operand" " r")))]
+  "TARGET_ZPSF"
+  "<uk>sub64 %0, %1, %2"
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+;; uk|k add|sub w|h
+(define_int_attr karth_insn
```

Comment:
U?K(sub|add)[HW] could be described similar to U?K(sub|add)<bits> + sign extension

Reply:
> U?K(sub|add)[HW] could be described similar to U?K(sub|add) + sign extension

Hi @marcfedorow , this is added in the latest update. Thanks for your feedback.